R/loo_moment_matching.R

```diff
@@ -0,0 +1,568 @@
+                                          is_method = is_method,
+                                          ...)
+      if (quantities_i$ki < ki) {
+        uparsi <- trans$upars
+        total_shift <- total_shift + trans$shift
+        total_scaling <- total_scaling * trans$scaling
+
+        lwi <- quantities_i$lwi
+        lwfi <- quantities_i$lwfi
+        ki <- quantities_i$ki
+        kfi <- quantities_i$kfi
+        log_liki <- quantities_i$log_liki
+        iterind <- iterind + 1
+        next
+      }
+
+      # 3. match means and covariances
+      if (cov) {
+        trans <- shift_and_cov(x, uparsi, lwi)
+        # gather updated quantities
+        quantities_i <- update_quantities_i(x, trans$upars,  i = i,
+                                            orig_log_prob = orig_log_prob,
+                                            log_prob_upars = log_prob_upars,
+                                            log_lik_upars = log_lik_upars,
+                                            r_effi = r_effi,
+                                            cores = cores,
+                                            is_method = is_method,
+                                            ...)
+
+        if (quantities_i$ki < ki) {
+          uparsi <- trans$upars
+          total_shift <- total_shift + trans$shift
+          total_mapping <- trans$mapping %*% total_mapping
+
+          lwi <- quantities_i$lwi
+          lwfi <- quantities_i$lwfi
+          ki <- quantities_i$ki
+          kfi <- quantities_i$kfi
+          log_liki <- quantities_i$log_liki
+          iterind <- iterind + 1
+          next
+        }
+      }
+      # none of the transformations improved khat
+      # so there is no need to try further
+      break
+    }
+
+    # transformations are now done
+    # if we don't do split transform, or
+    # if no transformations were successful
+    # stop and collect values
+    if (split && (iterind > 1)) {
+      # compute split transformation
+      split_obj <- split_mmloo(
+        x, upars, cov, total_shift, total_scaling, total_mapping, i,
+        log_prob_upars = log_prob_upars, log_lik_upars = log_lik_upars,
+        cores = cores, r_effi = r_effi, is_method = is_method
+      )
+      log_liki <- split_obj$log_liki
+      lwi <- split_obj$lwi
+      lwfi <- split_obj$lwfi
+      r_effi <- split_obj$r_effi
+    }
+    else {
+      dim(log_liki) <- c(S_per_chain, N_chains, 1)
+      r_effi <- loo::relative_eff(exp(log_liki), cores = cores)
+      dim(log_liki) <- NULL
+    }
+
+
+    # pointwise estimates
+    # elpd_loo
+    elpd_loo_i <- matrixStats::logSumExp(log_liki + lwi)
+    loo$pointwise[i, "elpd_loo"] <- elpd_loo_i
+    # p_loo
+    lpd <- matrixStats::logSumExp(log_liki) - log(length(log_liki))
+    loo$pointwise[i, "p_loo"] <- lpd - elpd_loo_i
+    # mcse_elpd_loo
+    loo$pointwise[i, "mcse_elpd_loo"] <- mcse_elpd(as.matrix(log_liki),as.matrix(lwi),
+                                     exp(elpd_loo_i), r_effi)
```

Comment:
Perhaps write this as 

```R
loo$pointwise[i, "mcse_elpd_loo"] <- mcse_elpd(
  as.matrix(log_liki), as.matrix(lwi), exp(elpd_loo_i), r_effi
)
```

with the arguments names made explit (as we currently have four unnamed arguments here)?

Reply:
done